theano/compile/function_module.py

```diff
@@ -483,6 +488,66 @@ def __copy__(self):
                 there.data = here.data
         return cpy
 
+    def copy(self, share_memory = False):
+        """
+        Copy this function. Copied function will have separated maker and fgraph
+        with original function. User can choose whether to separate storage by 
+        changing the share_memory arguments
+        ---------------------
+        Params:
+            share_memory -- { boolean } Default is False. When True, two function
+            share intermediate storages(storages except input and output storages)
+        ---------------------
+        Returns:
+            func -- Copied theano.Function
+        """
+
+        if not share_memory:
+            return self.__copy__()
+        else:
+            # copy SymbolocKits
+            ins, outs = copy.deepcopy([self.maker.inputs, self.maker.outputs])
```

Comment:
Instead of redoing all this, what about changing FunctionGraph.clone_get_equiv() to have a new parameter add_features that default to True and you set it to False? I think it would do what you need.

Another probably better option would be to add an attribute to Feature that need a new instance before being used in another fgraph. I would call it something like reinstanciate_on_reattach=True. You could set it to the DestroyHandler that caused you problems.